test/server/server_test.cc

```diff
@@ -134,6 +134,93 @@ TEST_F(RunHelperTest, ShutdownBeforeInitManagerInit) {
   target.ready();
 }
 
+class InitializingInitManager : public Init::ManagerImpl {
+public:
+  InitializingInitManager(absl::string_view name) : Init::ManagerImpl(name) {}
+
+  State state() const override { return State::Initializing; }
+};
+
+class InitializingInstanceImpl : public InstanceImpl {
+private:
+  InitializingInitManager init_manager_{"Server"};
+
+public:
+  InitializingInstanceImpl(const Options& options, Event::TimeSystem& time_system,
+                           Network::Address::InstanceConstSharedPtr local_address,
+                           ListenerHooks& hooks, HotRestart& restarter, Stats::StoreRoot& store,
+                           Thread::BasicLockable& access_log_lock,
+                           ComponentFactory& component_factory,
+                           Runtime::RandomGeneratorPtr&& random_generator,
+                           ThreadLocal::Instance& tls, Thread::ThreadFactory& thread_factory,
+                           Filesystem::Instance& file_system,
+                           std::unique_ptr<ProcessContext> process_context)
+      : InstanceImpl(options, time_system, local_address, hooks, restarter, store, access_log_lock,
+                     component_factory, std::move(random_generator), tls, thread_factory,
+                     file_system, std::move(process_context)) {}
+
+  Init::Manager& initManager() override { return init_manager_; }
+};
+
+// Class creates minimally viable server instance for testing and the initManager managed by it is
+// stuck in Initializing.
+class InitializingServerInstanceImplTest
+    : public testing::TestWithParam<Network::Address::IpVersion> {
+protected:
+  InitializingServerInstanceImplTest() : version_(GetParam()) {}
+
+  void initialize(const std::string& bootstrap_path) {
+    if (bootstrap_path.empty()) {
+      options_.config_path_ = TestEnvironment::temporaryFileSubstitute(
+          "test/config/integration/server.json", {{"upstream_0", 0}, {"upstream_1", 0}}, version_);
+    } else {
+      options_.config_path_ = TestEnvironment::temporaryFileSubstitute(
+          bootstrap_path, {{"upstream_0", 0}, {"upstream_1", 0}}, version_);
+    }
+    thread_local_ = std::make_unique<ThreadLocal::InstanceImpl>();
+    if (process_object_ != nullptr) {
```

Comment:
It seems like a bunch of this is not needed for the test? Instead of copying it all, can you refactor in same way to share the code?

Reply:
Yup. We can do that. Recent commit has that refactoring.